qiskit/aqua/operators/primitive_ops/matrix_op.py

```diff
@@ -95,10 +95,9 @@ def adjoint(self) -> OperatorBase:
                         coeff=np.conj(self.coeff))
 
     def equals(self, other: OperatorBase) -> bool:
-        if not isinstance(other, MatrixOp) or not self.coeff == other.coeff:
+        if not isinstance(other, MatrixOp):
             return False
-
-        return self.primitive == other.primitive
+        return self.coeff * self.primitive == other.coeff * other.primitive
```

Comment:
I think the right solution to this is more complex than you have given coeff can include things like ParameterExpression. You can see an aspect of this in add().

Also, I think that if we cannot do the comparison we should be raising NotIimplemented rather than returning False such as the test for other being MatrixOp or not. I know this is from the original code, and not part of your change as such, but its something that we need to re-consider so I am just noting it here.

Reply:
Thanks for your review!

Indeed! Thanks for the flag! I added basic expression comparison. Not really sure how to handle the issue for this type of parameter (like `"2 * y" * X == "y" * (2.X)`) as I couldn't find an easy way in `sympy` to get the possible factorisation between 2 expressions...

I understand your point and tried it but this caused the `reduce` function to fail on this line https://github.com/Qiskit/qiskit-aqua/blob/95d89596cb4670f05530aad2ff77724fe306118a/qiskit/aqua/operators/list_ops/summed_op.py#L97
Based on this, I think it actually makes sense to return `False` for not implemented comparisons.

Reply:
I imagined that returning NotImplemented may take adjustments elsewhere - hence it was more of a note. Its something that needs further investigation across the board outside of the scope of this PR.

We ideally ought to have a unit test that goes through the path of Parameter checking where it returns False .

I think the logic you have now is ok, thanks for the contribution. I will note the the primitive can still be different types but the Operator from Terra quantum_info, one of the types, supports multiply so I think its all good. A mix of different primitive types will continue to return False, as indeed it did before (even if they may really represent the same thing).

Reply:
I see! If you eventually decide to go with this `NotImplementedError` change and if needed, I can help with that :) 

I added a unit test checking `equals` returns `False` on different coefficient types (`expr * z != 2 * z`). Is it what you meant or you wanted a unit test for `expr1 * z != expr2 * z`?

Thank you for your reviews. You're welcome, my pleasure to help :) Yes indeed, I have tried to learn from your previous comment and looked at the types hints this time. From what I have seen, Operator implements `_multiply()` which is matched to `*` by its parent class `BaseOperator` so this part should indeed be fine. But when removing the type ignore comment, mypy was still failing because multiplications between `complex` and `list` are not supported. I copied the type fix from the `add` function just above the `equals` one https://github.com/Qiskit/qiskit-aqua/blob/95d89596cb4670f05530aad2ff77724fe306118a/qiskit/aqua/operators/primitive_ops/matrix_op.py#L84-L88

I also agree on your last point. Though this might not be the perfect behavior right now, I think this change should live in another issue/PR focused on converting between possible primitive types. I can also help on that if needed though I might need details/links on the expected behavior when comparing different representations.

Reply:
Thanks, the unit test you added was indeed what I had in mind. 

The primitive type in this derived class MatrixOp is more constrained than the types of its base class. list as you see is converted an Operator so the type ignore ends up being needed for what mypy sees as the full set types which are in fact less here. Thanks for looking through that though and checking too.